resolver/dns/dns_resolver_test.go

```diff
@@ -1126,3 +1142,59 @@ func TestCustomAuthority(t *testing.T) {
 		}
 	}
 }
+
+func TestRateLimitedResolve(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	// Create a new testResolver{} for this test because we want the exact count
+	// of the number of times the resolver was invoked.
+	nc := replaceNetFunc()
+	defer nc()
+
+	target := "foo.bar.com"
+	b := NewBuilder()
+	cc := &testClientConn{target: target}
+	r, err := b.Build(resolver.Target{Endpoint: target}, cc, resolver.BuildOption{})
+	if err != nil {
+		t.Fatalf("resolver.Build() returned error: %v\n", err)
+	}
+	dnsR, ok := r.(*dnsResolver)
+	if !ok {
+		t.Fatalf("resolver.Build() returned unexpected type: %T\n", dnsR)
+	}
+	tr, ok := dnsR.resolver.(*testResolver)
+	if !ok {
+		t.Fatalf("delegate resolver returned unexpected type: %T\n", tr)
+	}
+
+	var wg sync.WaitGroup
+	for i := 0; i < 20; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			r.ResolveNow(resolver.ResolveNowOption{})
+		}()
+	}
+	wg.Wait()
+
+	wantAddrs := []resolver.Address{{Addr: "1.2.3.4" + colonDefaultPort}, {Addr: "5.6.7.8" + colonDefaultPort}}
+	var gotAddrs []resolver.Address
+	var cnt int
+	for {
+		gotAddrs, cnt = cc.getAddress()
+		if cnt > 0 {
+			break
+		}
+		time.Sleep(time.Millisecond)
+	}
+	if !reflect.DeepEqual(gotAddrs, wantAddrs) {
+		t.Errorf("Resolved addresses of target: %q = %+v, want %+v\n", target, gotAddrs, wantAddrs)
+	}
+
+	gotResolveCnt := tr.cnt[target]
+	wantResolveCnt := 1
```

Comment:
We should also verify that we get two resolves eventually, but no more than two.

Maybe shorten the delay to avoid the test taking longer while it's just sleeping?

Reply:
Done. Let me know if you feel the `time.Duration` used in this test is too short/long.